File: tests/fixtures/swift/Buggy.swift
// EXPECTED FINDINGS:
// - line 17: swift-hardcoded-api-key — security — Hardcoded apiKey literal
// - line 19: swift-weak-crypto-md5 — security — Insecure.MD5 hash
// - line 23: swift-http-cleartext-url — security — URL with cleartext http://
// - line 27: swift-print-sensitive — security — print() statement
// - line 31: swift-force-unwrap — bug — Force unwrap with !
// - line 37: swift-empty-catch — bug — Empty catch block
// - line 47: swift-fatalError-prod — logic — fatalError in production code
// - line 50: swift-todo-comment — quality — TODO comment

import CryptoKit
import Foundation

class Buggy {

    func sample() {
        let apiKey = "live-prod-secret-9999"

        _ = Insecure.MD5.hash(data: Data())
    }

    func makeUrl() -> URL? {
        return URL(string: "http://example.com/api")
    }

    func debugTrace(value: String) {
        print(value)
    }

    func firstChar(s: String?) -> Character {
        return s!.first!
    }

    func safelyParse(_ s: String) -> Int? {
        do {
            return try parseInt(s)
        } catch {
        }
        return nil
    }

    func parseInt(_ s: String) throws -> Int {
        return Int(s) ?? 0
    }

    func unreachable() -> Never {
        fatalError("not implemented")
    }

    // TODO: refactor this entire module before shipping
    func placeholder() {}
}

File: tests/fixtures/dart/buggy.dart
// EXPECTED FINDINGS:
// - line 17: dart-hardcoded-api-key — security — Hardcoded apiKey constant
// - line 19: dart-weak-random — security — Random() not cryptographically secure
// - line 22: dart-http-cleartext-url — security — Cleartext HTTP URL string literal
// - line 26: dart-runtime-process-run — security — Process.run with variable command
// - line 30: dart-shared-prefs-secret — security — SharedPreferences token storage
// - line 34: dart-print-sensitive — security — print() statement
// - line 38: dart-null-assertion — bug — Null-assertion operator !
// - line 44: dart-empty-catch — bug — Empty catch block
// - line 48: dart-late-without-init — bug — late variable without initializer
// - line 50: dart-todo-comment — quality — TODO comment

import 'dart:io';
import 'dart:math';

class Buggy {
  static const apiKey = "live-prod-secret-9999";

  int randomDigit() => Random().nextInt(10);

  String apiUrl() {
    return "http://api.example.com/v1/data";
  }

  Future<ProcessResult> runCmd(String cmd, List<String> args) {
    return Process.run(cmd, args);
  }

  void persistToken(dynamic prefs, String value) {
    prefs.setString("token", value);
  }

  void debugTrace(String value) {
    print("trace: " + value);
  }

  String firstChar(String? s) {
    return s!.substring(0, 1);
  }

  int safelyParse(String s) {
    try {
      return int.parse(s);
    } catch (e) {}
    return 0;
  }

  late String userId;

  // TODO: implement caching layer before release
  void placeholder() {}
}
